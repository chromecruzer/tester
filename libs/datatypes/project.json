{
  "name": "datatypes",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/datatypes/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/datatypes/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/datatypes"],
      "options": {
        "jestConfig": "libs/datatypes/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
