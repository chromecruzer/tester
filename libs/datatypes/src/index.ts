export * from './datatypes';
export * from './IolRecord';
export * from './CustomerRecord';
export * from './CustomerContactRecord';
export * from './ProductRecord';
export * from './EmployeeRecord';        // added by shyam for testing 
export * from './EmployeeRecordDataMappings'   //
export * from './UserRecord1'  //
export * from './UserRecordDM' // 
export * from './ConsignmentRecord';
export * from './CustomerDataMapping';
export * from './CustomerContactRecord';
export * from './CustomerContactDataMapping';
export * from './IolDataMapping';
export * from './notifications';
export * from './ProductDataMapping';
export * from './SalesRepDataMapping';
export * from './SalesRepRecord';
export * from './SalesMappingSquared';
export * from './SalesMappingRecord';
export * from './SalesJumpRecord';
export * from './UploadHeaders';
export * from './AuditRecord';
export * from './SearchSignature';
export * from './EmailSettings';
//export * from './UserRecord';
import BuildUrl from "./BuildUrl";
import {ExpirationCalculations} from "./ExpirationCalculations";
import ItemMatchStatus from './ItemMatchStatus';
import SearchSignature from "./SearchSignature";
export {BuildUrl, ExpirationCalculations, ItemMatchStatus, SearchSignature};
