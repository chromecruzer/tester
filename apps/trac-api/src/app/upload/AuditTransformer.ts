import {UploadTransformerA} from "./UploadTransformerA";
import _ from "lodash";
import {DateTime} from "luxon";
import {AuditItemRecord, ItemMatchStatus, tracDateFormat, uploadHeaders} from "@trac/datatypes";
import util from "util";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export default class AuditTransformer extends UploadTransformerA {
  public convertSpreadsheetRecords(uploadedData: any[]) {
    return uploadedData;
  }
  constructor(storeFactory, schemaName, auditItemTableName,
              private auditTableName, private consignmentSchemaAndTableName, private matchStatus: ItemMatchStatus) {
    super(storeFactory, schemaName, auditItemTableName);
  }

  public async transform(uploadedData: string[], client, token: string, form) {
    const auditRecord = await this.extractLocation(form, client);
    // console.log(`matching the list ${dump(uploadedData)}`);
    if (!_.isEmpty(uploadedData)) {
      auditRecord.items = await this.matchExisting(auditRecord, uploadedData, client);
      // let lotNumbers = _.map(auditRecord.items, i => i.lot);
      // console.log(`audit items found original ${lotNumbers.length} records versus ${_.uniq(lotNumbers).length} unique records`)
      auditRecord.items = _.concat(auditRecord.items, await this.findMissing(auditRecord, uploadedData, client));
      // lotNumbers = _.map(auditRecord.items, i => i.lot);
      // console.log(`audit items with missing found original ${lotNumbers.length} records versus ${_.uniq(lotNumbers).length} unique records`)
    }
    // console.log(`auditRecord ${dump(auditRecord)} and token ${token}`)
    await this.storeFactory.create(token, auditRecord);
  }

  private async findMissing(auditRecord, uploadedData, client) {
    const matchlist = _.map(uploadedData, lot => `'${lot}'`);
    const sql = `SELECT * FROM ${(this.consignmentSchemaAndTableName)}
WHERE customer_id='${auditRecord.location.location_code}'
AND lot NOT IN (${_.join(matchlist,',')});`
    const missing = await client.query(sql)
      .catch(err => {
        throw new Error(`Query '${sql}' failed because ${err.message}`);
      });
    return _.map(missing.rows, m => ({
      lot: m.lot,
      consignment_uuid: m.uuid,
      item: m.item,
      quantity: m.quantity,
      family: m.description4,
      description: m.description,
      expire_date: m.expire_date,
      audit_match: 'Missing',
      last_changed_date: DateTime.utc().toJSDate(),
    }));
  }

  private async matchExisting(auditRecord, uploadedData, client) {
    const matchlist = _.map(uploadedData, lot => `'${lot}'`);
    const now = DateTime.utc().toJSDate();
    const sql = `SELECT * FROM ${(this.consignmentSchemaAndTableName)}
WHERE lot IN (${_.join(matchlist,',')});`
    // console.log(`query for matchlist ${sql}`)
    const matches = await client.query(sql)
      .catch(err => {
        throw new Error(`Query '${sql}' failed because ${err.message}`);
      });

    return _.map(uploadedData, numberColumn => {
      const found = _.find(matches.rows, m => m.lot === numberColumn);
      if (found) {
        const result: AuditItemRecord = {
          lot: numberColumn,
          item: found.item,
          consignment_uuid: found.uuid,
          family: found.description4,
          description: found.description,
          expire_date: found.expire_date,
          audit_match: this.matchStatus.match(auditRecord.location.location_code, found),
          quantity: found.quantity,
          last_changed_date: now
        };
        // console.log(`result match checks ${result.audit_match}
        // has ${result.audit_match === 'Found In Other Location' ? 'moved' : 'not moved'}
        // location ${found.customer_id}`);
        if (result.audit_match === 'Found In Other Location') {
          result.consignment_location = found.customer_id;
          result.consignment_warehouse = found.warehouse;
        }
        return result;
      }
      return {
        lot: numberColumn,
        audit_match: 'No Match',
        last_changed_date: now,
      };
    });
  }

  private async extractLocation(form, client) {
    const missing = [];
    ['location_code', 'scan_date', 'user', 'scanner', 'is_set'].forEach(field => {
      if (_.has(form, field)) {
        return;
      }
      missing.push(field);
    });
    if (!_.isEmpty(missing)) {
      throw new Error(`Upload audit form is missing ${_.join(missing, ',')}`);
    }
    const locationRecord = {
      received_date: DateTime.utc().toJSDate(),
      location_code: form[uploadHeaders.locationField],
      scan_date: form[uploadHeaders.scanDateField] ?
        DateTime.fromFormat(form[uploadHeaders.scanDateField], tracDateFormat).toUTC().toJSDate() :
        null,
      close_date: null,
      status: form[uploadHeaders.isSetField] === 'AUDIT' ? 'OPEN' : 'CHECKIN',
      auditor: form[uploadHeaders.userField],
      scanner: form[uploadHeaders.scannerField],
    }
    const sql = `SELECT * FROM ${this.schemaName}.${this.auditTableName}
WHERE location_code='${form[uploadHeaders.locationField]}'`;
    const audits = await client.query(sql)
      .catch(err => {
        throw new Error(`Query '${sql}' because ${err.message}`)
      });
    const found = _.find(audits.rows, a => {
      // console.log(`checking audit ${dump(a)}`);
      return a.scan_date === locationRecord.scan_date;
    });
    if (!_.isEmpty(found)) {
      throw Error(`Audit for location code ${locationRecord.location_code}
      scanned on ${locationRecord.scan_date} is already in trac`);
    }
    return {
      location: locationRecord,
      items: [],
      notes: [],
    };
  }
}
