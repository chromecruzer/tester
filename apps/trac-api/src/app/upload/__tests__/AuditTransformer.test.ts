import AuditTransformer from "../AuditTransformer";
import {ItemMatchStatus} from "@trac/datatypes";
import {DateTime} from "luxon";
import _ from "lodash";

describe('AuditTransformer', () => {
  let clientMock, mockStoreFactory, transformer, trueMatchesAudit, trueMatchesConsignment, misMatchedAudit,
    misMatchedConsignment, missingConsignment, nonmatchedAudit, auditLocation, consignmentLocation;
  const cleanupAuditResultsNow = () => {
    const audit = mockStoreFactory.create.mock.calls[0][1];
    const now = 'the time is now';
    audit.location.received_date = now;
    _.forEach(audit.items, i => {
      i.last_changed_date = now;
    });
    _.forEach(audit.notes, n => {
      n.date_created = now;
    });
    return audit
  };
  beforeEach(() => {
    clientMock = {
      query: jest.fn(),
    };
    mockStoreFactory = {
      create: jest.fn()
    }
    auditLocation = 'audit location';
    consignmentLocation = 'consignment location';
    clientMock.query.mockImplementation(() => Promise.resolve());
    transformer = new AuditTransformer(mockStoreFactory, 'schema name',
      'audit item table name', 'audit table name',
      'consignment schema & table name', new ItemMatchStatus());
    trueMatchesAudit = [
      'true match 1',
      'true match 2',
    ];
    trueMatchesConsignment = [
      {
        uuid: 'consignment uuid 1',
        lot: 'true match 1',
        customer_id: auditLocation,
        description4: 'dynasty',
        description: 'vixen',
        item: 'i am an item',
        quantity: 1,
        expire_date: DateTime.fromObject({year: 2023, month: 3, day: 15})
      },
      {
        uuid: 'consignment uuid 2',
        lot: 'true match 2',
        customer_id: auditLocation,
        description4: 'dallas',
        description: 'larry',
        item: 'i am an item',
        quantity: 1,
        expire_date: DateTime.fromObject({year: 2023, month: 3, day: 15})
      },
    ];

    misMatchedAudit = [
'missed match 1',
      'missed match 2',
    ];
    misMatchedConsignment = [
      {
        uuid: 'consignment uuid 3',
        lot: 'missed match 1',
        customer_id: consignmentLocation,
        description4: 'dynasty',
        description: 'vixen',
        item: 'i am an item',
        quantity: 1,
        expire_date: DateTime.fromObject({year: 2023, month: 3, day: 15})
      },
      {
        uuid: 'consignment uuid 4',
        lot: 'missed match 2',
        customer_id: consignmentLocation,
        description4: 'dallas',
        description: 'larry',
        item: 'i am an item',
        quantity: 1,
        expire_date: DateTime.fromObject({year: 2023, month: 3, day: 15})
      },
    ];
    nonmatchedAudit = [
 'no match 1',
      'no match 2',
    ];
    missingConsignment = [
      {
        uuid: 'consignment uuid 5',
        lot: 'missed scan 1',
        customer_id: consignmentLocation,
        description4: 'dynasty',
        description: 'vixen',
        item: 'i am an item',
        quantity: 1,
        expire_date: DateTime.fromObject({year: 2023, month: 3, day: 15})
      },
      {
        uuid: 'consignment uuid 6',
        lot: 'missed scan 2',
        customer_id: consignmentLocation,
        description4: 'dallas',
        description: 'larry',
        item: 'i am an item',
        quantity: 1,
        expire_date: DateTime.fromObject({year: 2023, month: 3, day: 15})
      },
    ];
  });
  describe('transform', () => {
    it('should find true matches', () => {
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: []}));
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: trueMatchesConsignment}));
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: []}));
      return transformer.transform(trueMatchesAudit, clientMock, 'I am a token', {
        user: 'User', scan_date: '10/23/2022',
        scanner: 'Sales Rep', location_code: auditLocation
      })
        .then(() => {
          expect(clientMock.query.mock.calls).toMatchSnapshot();
          expect(cleanupAuditResultsNow()).toMatchSnapshot();
        });
    });
    it('should find lenses that have been moved', () => {
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: []}));
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: misMatchedConsignment}));
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: []}));
      return transformer.transform(misMatchedAudit, clientMock, 'I am a token', {
        user: 'User', scan_date: '10/23/2022',
        scanner: 'Sales Rep', location_code: auditLocation
      })
        .then(() => {
          expect(cleanupAuditResultsNow()).toMatchSnapshot();
        });
    });
    it('should report audit entries that could not be matched', () => {
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: []}));
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: []}));
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: []}));
      return transformer.transform(nonmatchedAudit, clientMock, 'I am a token', {
        user: 'User', scan_date: '10/23/2022',
        scanner: 'Sales Rep', location_code: auditLocation
      })
        .then(() => {
          expect(cleanupAuditResultsNow()).toMatchSnapshot();
        });
    });
    it('should report audit entries that are missing from the audit scan', () => {
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: []}));
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: trueMatchesConsignment}));
      clientMock.query.mockImplementationOnce(() => Promise.resolve({rows: missingConsignment}));
      return transformer.transform(trueMatchesAudit, clientMock, 'I am a token', {
        user: 'User', scan_date: '10/23/2022',
        scanner: 'Sales Rep', location_code: auditLocation
      })
        .then(() => {
          expect(cleanupAuditResultsNow()).toMatchSnapshot();
        });
    });
  });
});
