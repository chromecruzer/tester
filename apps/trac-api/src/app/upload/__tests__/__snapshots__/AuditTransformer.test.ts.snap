// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuditTransformer transform should find lenses that have been moved 1`] = `
Object {
  "items": Array [
    Object {
      "audit_match": "Found In Other Location",
      "consignment_location": "consignment location",
      "consignment_uuid": "consignment uuid 3",
      "description": "vixen",
      "expire_date": "2023-03-15T00:00:00.000-05:00",
      "family": "dynasty",
      "item": "i am an item",
      "last_changed_date": "the time is now",
      "lot": "missed match 1",
      "quantity": 1,
    },
    Object {
      "audit_match": "Found In Other Location",
      "consignment_location": "consignment location",
      "consignment_uuid": "consignment uuid 4",
      "description": "larry",
      "expire_date": "2023-03-15T00:00:00.000-05:00",
      "family": "dallas",
      "item": "i am an item",
      "last_changed_date": "the time is now",
      "lot": "missed match 2",
      "quantity": 1,
    },
  ],
  "location": Object {
    "auditor": "User",
    "close_date": null,
    "location_code": "audit location",
    "received_date": "the time is now",
    "scan_date": 2022-10-23T05:00:00.000Z,
    "scanner": "Sales Rep",
    "status": "OPEN",
  },
  "notes": Array [],
}
`;

exports[`AuditTransformer transform should find true matches 1`] = `
Array [
  Array [
    "SELECT * FROM schema name.audit table name
WHERE location_code='audit location'",
  ],
  Array [
    "SELECT * FROM consignment schema & table name
WHERE lot IN ('true match 1','true match 2');",
  ],
  Array [
    "SELECT * FROM consignment schema & table name
WHERE customer_id='audit location'
AND lot NOT IN ('true match 1','true match 2');",
  ],
]
`;

exports[`AuditTransformer transform should find true matches 2`] = `
Object {
  "items": Array [
    Object {
      "audit_match": "True Match",
      "consignment_uuid": "consignment uuid 1",
      "description": "vixen",
      "expire_date": "2023-03-15T00:00:00.000-05:00",
      "family": "dynasty",
      "item": "i am an item",
      "last_changed_date": "the time is now",
      "lot": "true match 1",
      "quantity": 1,
    },
    Object {
      "audit_match": "True Match",
      "consignment_uuid": "consignment uuid 2",
      "description": "larry",
      "expire_date": "2023-03-15T00:00:00.000-05:00",
      "family": "dallas",
      "item": "i am an item",
      "last_changed_date": "the time is now",
      "lot": "true match 2",
      "quantity": 1,
    },
  ],
  "location": Object {
    "auditor": "User",
    "close_date": null,
    "location_code": "audit location",
    "received_date": "the time is now",
    "scan_date": 2022-10-23T05:00:00.000Z,
    "scanner": "Sales Rep",
    "status": "OPEN",
  },
  "notes": Array [],
}
`;

exports[`AuditTransformer transform should report audit entries that are missing from the audit scan 1`] = `
Object {
  "items": Array [
    Object {
      "audit_match": "True Match",
      "consignment_uuid": "consignment uuid 1",
      "description": "vixen",
      "expire_date": "2023-03-15T00:00:00.000-05:00",
      "family": "dynasty",
      "item": "i am an item",
      "last_changed_date": "the time is now",
      "lot": "true match 1",
      "quantity": 1,
    },
    Object {
      "audit_match": "True Match",
      "consignment_uuid": "consignment uuid 2",
      "description": "larry",
      "expire_date": "2023-03-15T00:00:00.000-05:00",
      "family": "dallas",
      "item": "i am an item",
      "last_changed_date": "the time is now",
      "lot": "true match 2",
      "quantity": 1,
    },
    Object {
      "audit_match": "Missing",
      "consignment_uuid": "consignment uuid 5",
      "description": "vixen",
      "expire_date": "2023-03-15T00:00:00.000-05:00",
      "family": "dynasty",
      "item": "i am an item",
      "last_changed_date": "the time is now",
      "lot": "missed scan 1",
      "quantity": 1,
    },
    Object {
      "audit_match": "Missing",
      "consignment_uuid": "consignment uuid 6",
      "description": "larry",
      "expire_date": "2023-03-15T00:00:00.000-05:00",
      "family": "dallas",
      "item": "i am an item",
      "last_changed_date": "the time is now",
      "lot": "missed scan 2",
      "quantity": 1,
    },
  ],
  "location": Object {
    "auditor": "User",
    "close_date": null,
    "location_code": "audit location",
    "received_date": "the time is now",
    "scan_date": 2022-10-23T05:00:00.000Z,
    "scanner": "Sales Rep",
    "status": "OPEN",
  },
  "notes": Array [],
}
`;

exports[`AuditTransformer transform should report audit entries that could not be matched 1`] = `
Object {
  "items": Array [
    Object {
      "audit_match": "No Match",
      "last_changed_date": "the time is now",
      "lot": "no match 1",
    },
    Object {
      "audit_match": "No Match",
      "last_changed_date": "the time is now",
      "lot": "no match 2",
    },
  ],
  "location": Object {
    "auditor": "User",
    "close_date": null,
    "location_code": "audit location",
    "received_date": "the time is now",
    "scan_date": 2022-10-23T05:00:00.000Z,
    "scanner": "Sales Rep",
    "status": "OPEN",
  },
  "notes": Array [],
}
`;
